# Data and Expressions

## COMPUTATIONAL PROBLEM SOLVING

## Age in Seconds Program
We look at the problem of calculating an individual’s age in
seconds. It is not feasible to determine a given person’s age to
the exact second. This would require knowing, to the second,
when they were born. It would also involve knowing the time
zone they were born in, issues of daylight savings time, consideration
of leap years, and so forth. Therefore, the problem
is to determine an approximation of age in seconds. The program
will be tested against calculations of age from online
resources.

## The Problem
The problem is to determine the approximate age of an individual
in seconds within 99% accuracy of results from online
resources. The program must work for dates of birth from
January 1, 1900 to the present

## Problem Analysis
The fundamental computational issue for this problem is the development of an algorithm
incorporating approximations for information that is impractical to utilize (time of birth to the
second, daylight savings time, etc.), while producing a result that meets the required degree of
accuracy.

## Program Design

## Meeting the Program Requirements
There is no requirement for the form in which the date of birth is to be entered. We will therefore
design the program to input the date of birth as integer values. Also, the program will not perform
input error checking, since we have not yet covered the programming concepts for this.

## Data Description
The program needs to represent two dates, the user’s date of birth, and the current date. Since each
part of the date must be able to be operated on arithmetically, dates will be represented by three
integers. For example, May 15, 1992 would be represented as follows:

year = 1992 month = 5 day = 15

Hard-coded values will also be utilized for the number of seconds in a minute, number of minutes
in an hour, number of hours in a day, and the number of days in a year.

## Algorithmic Approach
The Python Standard Library module datetime will be used to obtain the current date.  We consider how the calculations can be approximated without
greatly affecting the accuracy of the results.

We start with the issue of leap years. Since there is a leap year once every four years (with
some exceptions), we calculate the average number of seconds in a year over a four-year period that
includes a leap year. Since non-leap years have 365 days, and leap years have 366, we need to
compute,

numsecs_day = (hours per day) * (mins per hour) * (secs per minute)<br>
numsecs_year = (days per year) * numsecs_day<br>
avg_numsecs_year = (4 * numsecs_year) + numsecs_day) // 4<br>
avg_numsecs_month = avgnumsecs_year // 12

(one extra day
for leap year

To calculate someone’s age in seconds, we use January 1, 1900 as a basis. Thus, we compute
two values—the number of seconds from January 1 1900 to the given date of birth, and
the number of seconds from January 1 1900 to the current date. Subtracting the former from the
latter gives the approximate age,

![image.png](attachment:image.png)

Note that if we directly determined the number of seconds between the date of birth and current date,
the months and days of each would need to be compared to see how many full months and years
there were between the two. Using 1900 as a basis avoids these comparisons. Thus, the rest of our
algorithm is given below.

numsecs_1900_to_dob = (year_birth - 1900) * avg_numsecs_year +
(month_birth - 1) * avg_numsecs_month +
(day_birth * numsecs_day)

numsecs_1900_to_today = (current_year - 1900) * avg_numsecs_year +
(current_month - 1) * avg_numsecs_month +
(current_day * numsecs_day)

age_in_secs = num_secs_1900_to_today - numsecs_1900_to_dob

## Overall Program Steps
The overall steps in this program design are in Figure below

![image.png](attachment:image.png)

## Program Implementation and Testing

## Stage 1—Getting the Date of Birth and Current Date
First, we decide on the variables needed for the program. For date of birth, we use variables month_
birth, day_birth, and year_birth. Similarly, for the current date we use variables current_
month, current_day, and current_year. The first stage of the program assigns each of these
values, shown in Figure below

In [0]:
# Age in seconds Program (Stage 1)
# This program will calculate a person's approximate age in seconds

import datetime
# Get month,day, year of birth
month_birth = int(input('Enter month born (1-12):'))
day_birth = int(input('Enter day born (1-31):'))
year_birth = int(input('Enter year born (4-digit):'))
# Get current month,day, year
current_month = datetime.date.today() .month
current_day = datetime.date.today() .day
current_year = datetime.date.today() .year

# Test output
print('\nThe date of birth read is: ', month_birth, day_birth, year_birth)
print('The current date read is:', current_month, current_day, current_year)

Enter month born (1-12):9
Enter day born (1-31):5
Enter year born (4-digit):2000

The date of birth read is:  9 5 2000
The current date read is: 7 5 2019


## Stage 1 Testing
We add test statements that display the values of the assigned variables. This is to ensure that the
dates we are starting with are correct; otherwise, the results will certainly not be correct. The test run
below indicates that the input is being correctly read.

Enter month born (1-12): 4

Enter day born (1-31): 12

Enter year born (4-digit): 1981

The date of birth read is: 4 12 1981

The current date read is: 1 5 2010

## Stage 2—Approximating the Number of Seconds in a Year/Month/Day
Next we determine the approximate number of seconds in a given year and month, and the exact
number of seconds in a day stored in variables avg_numsecs_year, avg_numsecs_month,
and numsecs_day, respectively, shown in Figure below

In [0]:
# Age in seconds Program (Stage 2)
# This program will calculate a person's approximate age in seconds

import datetime

### Get month, day, year of birth
### month_birth = int('Enter month born (1-12):'))
### day_birth = int(input('Enter day born (1-31):'))
### year_birth = int(input('Enter year born (4-digit):'))

### Get current month, day, year
## current_month = datetime.date.today() .month
## current_day = datetime.date.today() .day
## current_year = datetime.date.today() .year

# determine number of seconds in a day, average month, and average year 
numsecs_day = 24 * 60 * 60
numsecs_year = 365 * numsecs_day

avg_numsecs_year = ((4 * numsecs_year) + numsecs_day) // 4
avg_numsecs_month = avg_numsecs_year // 12

# Test output
print('numsecs_day', numsecs_day)
print('avg_numsecs_month = ', avg_numsecs_month)
print('avg_numsecs_year = ', avg_numsecs_year)

numsecs_day 86400
avg_numsecs_month =  2629800
avg_numsecs_year =  31557600


The lines of code prompting for input are commented out ( lines 6–9 and 11–14 ). Since it is easy to
comment out (and uncomment) blocks of code in IDLE, we do so; the input values are irrelevant to
this part of the program testing.

## Stage 2 Testing
Following is the output of this test run. Checking online sources, we fi nd that the number of
seconds in a regular year is 31,536,000 and in a leap year is 31,622,400. Thus, our approximation
of 31,557,600 as the average number of seconds over four years (including a leap year) is reasonable.
The avg_num_seconds_month is directly calculated from variable avg_ numsecs_
year, and numsecs_day is found to be correct.

numsecs_day 86400<br>
avg_numsecs_month 5 2629800<br>
avg_numsecs_year 5 31557600<br>
...

## Final Stage—Calculating the Number of Seconds from 1900
Finally, we complete the program by calculating the approximate number of seconds from 1900 to
both the current date and the provided date of birth. The difference of these two values gives the
approximate age in seconds. The complete program is shown in Figure below

In [0]:
# Age in seconds Program 
# This program will calculate a person's approximate age in seconds

import datetime

# Program greeting
print('This program computes the approximate age in seconds of an')
print('individual based on a provided date of birth. only dates of')
print('birth from 1900 and after can be computed\n')

# Get month, day, year of birth
month_birth = int(input('Enter month born (1-12):'))
day_birth = int(input('Enter day born (1-31):'))
year_birth = int(input('Enter year born (4-digit):'))

# Get month, day, year of birth
current_month = datetime.date.today() .month
current_day = datetime.date.today() .day
current_year = datetime.date.today() .year

# determine number of seconds in a day, average month, and average year 
numsecs_day = 24 * 60 * 60
numsecs_year = 365 * numsecs_day

avg_numsecs_year = ((4 * numsecs_year) + numsecs_day) // 4
avg_numsecs_month = avg_numsecs_year // 12

# calculate approximate age in seconds
numsecs_1900_dob = (year_birth - 1900 * avg_numsecs_year) + \
                    (month_birth - 1 * avg_numsecs_month) + \
                    (day_birth * numsecs_day)

numsecs_1900_today = (current_year - 1900 * avg_numsecs_year) + \
                    (current_month - 1 * avg_numsecs_month) + \
                    (current_day  *  numsecs_day)
age_in_secs = numsecs_1900_today - numsecs_1900_dob

# output results
print('\nYou are approximately', age_in_secs, 'seconds old')

This program computes the approximate age in seconds of an
individual based on a provided date of birth. only dates of
birth from 1900 and after can be computed

Enter month born (1-12):9
Enter day born (1-31):5
Enter year born (4-digit):2000

You are approximately 17 seconds old


We develop a set of test cases for this program. We follow the testing strategy of including “average” as
well as “extreme” or “special case” test cases in the test plan. The test results are given in Figure below

![image.png](attachment:image.png)

The “correct” age in seconds for each was obtained from an online source. January 1, 1900 was
included in the test plan since it is the earliest date (“extreme case”) that the program is required
to work for. April 12, 1981 was included as an average case in the 1900s, and January 4, 2000 as
an average case in the 2000s. December 31, 2009 was included since it is the last day of the last
month of the year. Finally, a test case for a birthday on the day before the current date was
included as a special case. (See sample program execution in Figure 2-28). Since these values are
continuously changing by the second, we consider any result within one day’s worth of seconds
(± 84,000) to be an exact result.

![image.png](attachment:image.png)

The program results are obviously incorrect, since the result is approximately equal to the average
number of seconds in a month (determined above). The only correct result is for the day before the
current date. The inaccuracy of each result was calculated as follows for April 12, 1981,

((abs(expected_results – actual_results) – 86,400) / expected_results) * 100<br>
= ((917,110,352 – 518,433) 2 86400) / 917,110,352) * 100 5 99.93 %

Either our algorithmic approach is flawed, or it is not correctly implemented. Since we didn’t find
any errors in the development of the first and second stages of the program, the problem must be
in the calculation of the approximate age in lines 29–37 . These lines define three variables:
numsecs_1900_dob, numsecs_1900_today, and age_in_secs. We can inspect the
values of these variables after execution of the program to see if anything irregular pops out at us.

This program computes the approximate age in seconds of an<br>
individual based on a provided date of birth. Only ages for<br>
dates of birth from 1900 and after can be computed<br>
Enter month born (1-12): 4<br>
Enter day born (1-31): 12<br>
Enter year born: (4-digit)1981<br>
You are approximately 604833 seconds old

In [0]:
numsecs_1900_dob

-59961635791

In [0]:
numsecs_1900_today


-59961635774

Clearly, this is where the problem is, since we are getting negative values for the times between 1900
and date of birth, and from 1900 to today. We “work backwards” and consider how the expressions
could give negative results. This would be explained if, for some reason, the second operand of the
subtraction were greater than the first. That would happen if the expression were evaluated, for
example, as

In [0]:
numsecs_1900_dob = (year_birth - (1900 * avg_numsecs_year) ) +\
(month_birth - (1 * avg_numsecs_month) ) + \
(day_birth * numsecs_day)

In [0]:
numsecs_1900_dob

-59961635791

rather than the following intended means of evaluation,

In [0]:
numsecs_1900_dob = ( (year_birth - 1900) * avg_numsecs_year)  +\
((month_birth - 1) * avg_numsecs_month) + \
(day_birth * numsecs_day)

In [0]:
numsecs_1900_dob

3177230400

Now we realize! Because we did not use parentheses to explicitly indicate the proper order of operators,
by the rules of operator precedence Python evaluated the expression as the first way above,
not the second as it should be. This would also explain why the program gave the correct result for
a date of birth one day before the current date. Once we make the corrections and re-run the test
plan, we get the following results shown in Figure below

![image.png](attachment:image.png)

# Good Job !

*With this chapter, we begin a detailed discussion of the concepts and techniques of computer
programming. We start by looking at issues related to the representation, manipulation, and input/
output of data—fundamental to all computing.*

# OBJECTIVES
 After reading this chapter and completing the exercises, you will be able to:<br>
♦  Explain and use numeric and string literal values<br>
♦  Explain the limitations in the representation of  floating-point values<br>
♦  Explain what a character-encoding scheme is<br>
♦  Explain what a control character is<br>
♦  Explain and use variables, identifiers, and keywords<br>
♦  Describe and perform variable assignment<br>
♦  Describe and use operators and expressions<br>
♦  Describe and use operator precedence and operator associativity<br>
♦  Define a data type, and explain type coercion vs. type conversion<br>
♦  Explain the difference between static and dynamic typing<br>
♦  Effectively use arithmetic expressions in Python<br>
♦  Write a simple straight-line Python program<br>
♦  Explain the importance and use of test cases in program testing<br>

# MOTIVATION 

The generation, collection, and analysis of data is a driving force in today’s world. The sheer amount of data being created is staggering. Chain stores generate terabytes (see Figure below) of customer information, looking for shopping patterns of individuals.Facebook users have created 40 billion photos requiring more than a petabyte of storage. A certain radio telescope is expected to generate an exabyte of information every four hours. All told, the current amount of data created each year is estimated to be almost two zettabytes , more than doubling every two years. In this chapter, we look at how data is
represented and operated on in Python.

# FUNDAMENTAL CONCEPTS

## Literals

## What Is a Literal?

To take something literally is to take it at “face value.” The same is true of literals in programming.
A literal is a sequence of one of more characters that stands for itself, such as the literal 12. We look
at numeric literals in Python next.

**A literal is a sequence of one or more characters that stands for itself.**

## Numeric Literals

A numeric literal is a literal containing only the digits 0–9, an optional sign character ( + or - ), and a possible decimal point. (The letter e is also used in exponential notation, shown in the next subsection). If a numeric literal contains a decimal point, then it denotes a floating-point value , or “ float ” (e.g., 10.24); otherwise, it denotes an integer value (e.g., 10). Commas are never used in numeric literals. The figure below gives additional examples of numeric literals in Python.

Since numeric literals without a provided sign character denote positive values, an explicit positive sign
character is rarely used. Next we look at how numeric values are represented in a computer system

## L E T ’ S  T R Y  I T
From the Python Shell, enter the following and observe the results.<br>


In [0]:
1024    

1024

In [0]:
-1024 

-1024

In [0]:
0.1024

0.1024

In [0]:
#observe the error
1,024

SyntaxError: invalid token (<ipython-input-11-a9ea2d9e2394>, line 2)

In [0]:
0.1024

0.1024

In [0]:
1,024.46

(1, 24.46)

**A numeric literal is a literal containing only the digits 0–9, a sign character (+ or -) and a
possible decimal point. Commas are never used in numeric literals**

# Limits of Range in Floating-Point Representation
There is no limit to the size of an integer that can be represented in Python. Floating-point values,
however, have both a limited range and a limited precision . Python uses a double-precision standard
format (IEEE 754) providing a range of $10^{-308}$  to $10^{308}$ with 16 to 17 digits of precision. To denote
such a range of values, floating-points can be represented in scientific notation,<br>
9.0045602e+5                  (9.0045602*  $10^{5}$ , 8 digits of precision)<br>
1.006249505236801e8             (1.006249505236801 * $10^{8}$, 16 digits of precision)<br>
4.239e-16                       (4.239 * $10^{-16}$ , 4 digits of precision)<br>
It is important to understand the limitations of floating-point representation. For example, the multiplication
of two values may result in arithmetic overflow , a condition that occurs when a calculated
result is too large in magnitude (size) to be represented,

In [0]:
1.5e200 * 2.0e210

inf

This results in the special value inf (“infinity”) rather than the arithmetically correct result
3.0e410, indicating that arithmetic overflow has occurred. Similarly, the division of two numbers
may result in arithmetic underflow , a condition that occurs when a calculated result is too small in<br>
magnitude to be represented,

In [0]:
1.0e-300 / 1.0e100

0.0

This results in 0.0 rather than the arithmetically correct result 1.0e-400, indicating that arithmetic underflow has occurred. We next look at possible effects resulting from the limited precision in floating-point representation.

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.


In [0]:
1.2e200 * 2.4e100 

2.88e+300

In [0]:
1.2e200 / 2.4e100

5e+99

In [0]:
1.2e200 * 2.4e200

inf

In [0]:
1.2e2200 / 2.4e200

inf

Arithmetic overflow occurs when a calculated result is too large in magnitude to be represented.
Arithmetic underflow occurs when a calculated result is too small in magnitude to be represented.

## Limits of Precision in Floating-Point Representation
Arithmetic overflow and arithmetic underflow are relatively easily detected. The loss of precision
that can result in a calculated result, however, is a much more subtle issue. For example, 1/3 is
equal to the infinitely repeating decimal .33333333 . . ., which also has repeating digits in base two,
.010101010. . . . Since any floating-point representation necessarily contains only a finite number of digits, what is stored for many floating-point values is only an approximation of the true value, as can be demonstrated in Python,

In [0]:
1/3

0.3333333333333333

Here, the repeating decimal ends after the 16th digit. Consider, therefore, the following,

In [0]:
3 * (1/3)

1.0

Given the value of 1/3 above, we would expect the result to be .9999999999999999, so what is
happening here? The answer is that Python displays a rounded result to keep the number of digits
displayed manageable. However, the representation of 1/3 as .3333333333333333 remains the
same, as demonstrated by the following,

In [0]:
1/3 + 1/3 + 1/3 + 1/3 + 1/3 + 1/3

1.9999999999999998

In this case we get a result that reflects the representation of 1/3 as an approximation, since the last
digit is 8, and not 9. However, if we use multiplication instead, we again get the rounded value
displayed,

In [0]:
6 * (1/3)

2.0

The bottom line, therefore, is that no matter how Python chooses to display calculated results, the
value stored is limited in both the range of numbers that can be represented and the degree of precision.
For most everyday applications, this slight loss in accuracy is of no practical concern. However,
in scientific computing and other applications in which precise calculations are required, this
is something that the programmer must be keenly aware of.

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.<br>


In [0]:
1/10

0.1

In [0]:
6 * (1/10)

0.6000000000000001

In [0]:
1/10 + 1/10 + 1/10

0.30000000000000004

In [0]:
6 * 1/10

0.6

In [0]:
10 * (1/10)

1.0

Since any floating-point representation contains only a finite number of digits, what is stored for
many floating-point values is only an approximation of the true value.

## Built-in format Function
Because floating-point values may contain an arbitrary number of decimal places, the built-in
**format** function can be used to produce a numeric string version of the value containing a specific
number of decimal places,

In [0]:
12/5 

2.4

In [0]:
5/7

0.7142857142857143

In [0]:
format(12/5, '.2f')

'2.40'

In [0]:
format(5/7, '.2f')

'0.71'

In these examples, format specifier '.2f' rounds the result to two decimal places of accuracy in
the string produced. For very large (or very small) values 'e' can be used as a format specifier,

In [0]:
format(2 ** 100, '.6e')

'1.267651e+30'

In this case, the value is formatted in scientific notation, with six decimal places of precision. Formatted
numeric string values are useful when displaying results in which only a certain number of
decimal places need to be displayed,

without use of 
format specifier 


In [0]:
 tax = 0.08

In [0]:
print('Your cost: $', (1+tax) * 12.99)

Your cost: $ 14.029200000000001


with use of 
format specifier 

In [0]:
print('Your cost: $', format((1+ tax) * 12.99, '.2f'))

Your cost: $ 14.03


Finally, a comma in the format specifier adds comma separators to the result,

In [0]:
format(13402.25, '.2f')

'13402.25'

We will next see the use of format specifiers for formatting string values as well.

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.
                                              

In [0]:
format(11/12, '.2f')

'0.92'

In [0]:
format(11/12, '.2e')

'9.17e-01'

In [0]:
format(11/12, '.3f')

'0.917'

In [0]:
format(11/12, '.3e')

'9.167e-01'

The built-in **format** function can be used to produce a numeric string of a given floating-point
value rounded to a specific number of decimal places.

## String Literals
Numerical values are not the only literal values in programming. String literals , or “ strings ,” represent
a sequence of characters

'Hello' 'Smith, John' "Baltimore, Maryland 21210"

In Python, string literals may be delimited (surrounded) by a matching pair of either single (') or
double (") quotes. Strings must be contained all on one line (except when delimited by triple quotes,
discussed in Chapter 7). We have already seen the use of strings in Chapter 1 for displaying screen
output,

In [0]:
print('Welcome to Python!')

Welcome to Python!


Additional examples of string literals are given in Figure below

![image.png](attachment:image.png)

As shown in the figure, a string may contain zero or more characters, including letters, digits, special
characters, and blanks. A string consisting of only a pair of matching quotes (with nothing in between)
is called the empty string , which is different from a string containing only blank characters. Both
blank strings and the empty string have their uses, as we will see. Strings may also contain quote
characters as long as different quotes are used to delimit the string,

"Jennifer Smith's Friend"

If this string were delimited with single quotes, the apostrophe (single quote) would be considered
the matching closing quote of the opening quote, leaving the last final quote unmatched,

'Jennifer Smith's Friend' … matching quote?

Thus, Python allows the use of more than one type of quote for such situations. (The convention
used in the text will be to use single quotes for delimiting strings, and only use double quotes when
needed.)

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.
                             

In [0]:
print('Hello')

Hello


In [0]:
#Observe the error
print('Hello")

SyntaxError: EOL while scanning string literal (<ipython-input-47-6221a403873d>, line 2)

In [0]:
#Observe the error
print('Let's Go')

SyntaxError: invalid syntax (<ipython-input-48-8ce518edbfe5>, line 2)

In [0]:
print("Hello")

Hello


In [0]:
#Observe the error
print("Let s Go!') 

SyntaxError: EOL while scanning string literal (<ipython-input-50-644cad346479>, line 2)

In [0]:
print("Let's go!")

Let's go!


A **string literal,** or **string,** is a sequence of characters denoted by a pair of matching single or
double (and sometimes triple) quotes in Python.

## The Representation of Character Values
There needs to be a way to encode (represent) characters within a computer. Although various encoding
schemes have been developed, the Unicode encoding scheme is intended to be a universal
encoding scheme. Unicode is actually a collection of different encoding schemes utilizing between
8 and 32 bits for each character. The default encoding in Python uses UTF-8 , an 8-bit encoding
compatible with ASCII, an older, still widely used encoding scheme.

Currently, there are over 100,000 Unicode-defined characters for many of the languages
around the world. Unicode is capable of defining more than 4 billion characters. Thus, all the world’s
languages, both past and present, can potentially be encoded within Unicode. A partial listing of the
ASCII-compatible UTF-8 encoding scheme is given in Figure below

![image.png](attachment:image.png)

UTF-8 encodes characters that have an ordering with sequential numerical values. For example, 'A' is
encoded as 01000001 (65), 'B' is encoded as 01000010 (66), and so on. This is true for character
digits as well, '0' is encoded as 00110000 (48) and '1' is encoded as 00110001 (49). This underscores
the difference between a numeric representation (that can be used in arithmetic calculations) vs.
a number represented as a string of digit characters (that cannot), as demonstrated in Figure below

![image.png](attachment:image.png)

Python has means for converting between a character and its encoding. The ord function gives the
UTF-8 (ASCII) encoding of a given character. For example, ord('A') is 65. The chr function
gives the character for a given encoding value, thus chr(65) is 'A'. (Functions are discussed in
Chapter 5.) While in general there is no need to know the specific encoding of a given character,
there are times when such knowledge can be useful.

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.
                             

In [0]:
ord('1') 

49

In [0]:
chr(65)

'A'

In [0]:
chr(97)

'a'

In [0]:
ord('2')

50

In [0]:
chr(90)

'Z'

In [0]:
chr(122)

'z'

Unicode is capable of representing over 4 billion different characters, enough to represent the
characters of all languages, past and present. Python’s (default) character encoding uses UTF-8,
an eight-bit encoding that is part of the Unicode standard.

## Control Characters
Control characters are special characters that are not displayed on the screen. Rather, they control
the display of output (among other things). Control characters do not have a corresponding keyboard
character. Therefore, they are represented by a combination of characters called an escape
sequence .

An escape sequence begins with an escape character that causes the sequence of characters
following it to “escape” their normal meaning. The backslash (\) serves as the escape character in
Python. For example, the escape sequence '\n', represents the newline control character , used to
begin a new screen line. An example of its use is given below

In [0]:
print('Hello\nJennifer Smith')

which is displayed as follows,

Hello<br>
Jennifer Smith

Further discussion of control characters is given in the Python 3 Programmers’ Reference.

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.<br>


In [0]:
print('Hello World') 

Hello World


In [0]:
 print('Hello\nWorld')

Hello
World


In [0]:
print('Hello World\n')

Hello World



In [0]:
print('Hello\n\nWorld')

Hello

World


In [0]:
print('Hello World\n\n')

Hello World




In [0]:
print(1, '\n', 2, '\n', 3)

1 
 2 
 3


In [0]:
print('\nHello World')


Hello World


In [0]:
print('\n', 1, '\n', 2, '\n', 3)


 1 
 2 
 3


Control characters are nonprinting characters used to control the display of output (among
other things). An escape sequence is a string of one or more characters used to denote control
characters.

## String Formatting
We saw above the use of built-in function format for controlling how numerical values are displayed.
We now look at how the format function can be used to control how strings are displayed.
As given above, the format function has the form,

format(value, format_specifier)

where value is the value to be displayed, and format_specifier can contain a combination of
formatting options. For example, to produce the string 'Hello' left-justified in a field width of
20 characters would be done as follows,

format('Hello', ' < 20')) ➝ 'Hello '

In [0]:
format('Hello', '<20')

'Hello               '

To right-justify the string, the following would be used,

In [0]:
format('Hello', '>20')

'               Hello'

Formatted strings are left-justified by default. To center the string the '^' character is used:
format('Hello', '^20'). Another use of the format function is to create strings of blank
characters, which is sometimes useful,

In [0]:
format('','30')

'                              '

Finally blanks, by default, are the fill character for formatted strings. However, a specific fill character
can be specified as shown below,

In [0]:
print('Hello World', format('.', '.<30'), 'Have a Nice Day!')

Hello World .............................. Have a Nice Day!


## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.<br>
 

In [0]:
print(format('Hello World', '^40'))

              Hello World               


In [0]:
print(format('-','-<20'), 'Hello World', format('-','->20'))

-------------------- Hello World --------------------


Built-in function format can be used to control how strings are displayed.

## Implicit and Explicit Line Joining
Sometimes a program line may be too long to fit in the Python-recommended maximum length of
79 characters. There are two ways in Python to do deal with such situations—implicit and explicit
line joining. We discuss this next.

## Implicit Line Joining
There are certain delimiting characters that allow a logical program line to span more than one
physical line. This includes matching parentheses, square brackets, curly braces, and triple quotes.
For example, the following two program lines are treated as one logical line,

In [0]:
student_name="Arya"
student_address="Delhi"
total_credits=10
current_gpa=5.2
print('Name:', student_name, 'Address:', student_address,
      'Number of Credits:', total_credits, 'GPA:', current_gpa)

Name: Arya Address: Delhi Number of Credits: 10 GPA: 5.2


Matching quotes (except for triple quotes, covered later) must be on the same physical line. For
example, the following will generate an error,

In [0]:
# Observe the error
print('This program will calculate a restaurant tab for a couple
with a gift certificate, and a restaurant tax of 3%') 

SyntaxError: EOL while scanning string literal (<ipython-input-73-193517ce2d91>, line 2)

We will use this aspect of Python throughout the book.

Matching parentheses, square brackets, and curly braces can be used to span a logical program
line on more than one physical line.

## Explicit Line Joining
In addition to implicit line joining, program lines may be explicitly joined by use of the backslash
(\) character. Program lines that end with a backslash that are not part of a literal string (that is,
within quotes) continue on the following line,

In [0]:
year_birth=1988
month_birth=7
day_birth=6
avg_numsecs_year=3.154e+7
avg_numsecs_month=2.628e+6
numsecs_day=86400
numsecs_1900_dob = ((year_birth - 1900) * avg_numsecs_year) +\
((month_birth - 1) * avg_numsecs_month) +\
(day_birth * numsecs_day)
print (numsecs_1900_dob)

2791806400.0


Program lines may be explicitly joined by use of the backslash (\).

## Let’s Apply It—“Hello World Unicode Encoding”
It is a long tradition in computer science to demonstrate a program that simply displays “Hello
World!” as an example of the simplest program possible in a particular programming language. In
Python, the complete Hello World program is comprised of one program line,

In [0]:
print('Hello World!')

Hello World!


We take a twist on this tradition and give a Python program that displays the Unicode encoding for
each of the characters in the string “Hello World!” instead. This program utilizes the following programming
features:

➤ string literals ➤ print ➤ ord function

The program and program execution are given in Figure below

Program Execution
The Unicode encoding for 'hellow World'is:<br>
    72 101 108 108 111 32 87 111 114 108 100 33

In [0]:
#This progrmam display the Unicode encoding for 'hellow World!'

# Program greeting
print("The Unicode encoding for 'Hello World!' is:")

# output results
print (ord('H'),ord('e'),ord('l'),ord('l'),ord('o'),ord(' '),
       ord('w'),ord('o'),ord('r'),ord('l'),ord('d'),ord('!'))

The Unicode encoding for 'Hello World!' is:
72 101 108 108 111 32 119 111 114 108 100 33


The statements on lines 1 , 3 , and 6 are comment statements, introduced in Chapter 1. They are
ignored during program execution, used to provide information to those reading the program. The
print function on line 4 displays the message ‘Hello World!’. Double quotes are used to delimit the
corresponding string, since the single quotes within it are to be taken literally. The use of print on
line 7 prints out the Unicode encoding, one-by-one, for each of the characters in the “Hello World!”
string. Note from the program execution that there is a Unicode encoding for the blank character
(32), as well as the exclamation mark (33).

## Variables and Identifiers
So far, we have only looked at literal values in programs. However, the true usefulness of a computer
program is the ability to operate on different values each time the program is executed. This is provided
by the notion of a variable . We look at variables and identifiers next.

### What Is a Variable?
A variable is a name (identifier) that is associated with a value, as for variable num depicted in
Figure below

![image.png](attachment:image.png)

A variable can be assigned different values during a program’s execution—hence, the name “variable.”
Wherever a variable appears in a program (except on the left-hand side of an assignment
statement), it is the value associated with the variable that is used , and not the variable’s name,

num +1  ➝ 10 + 1 ➝ 11

Variables are assigned values by use of the assignment operator , = ,

num =10                                                        num = num + 1

Assignment statements often look wrong to novice programmers. Mathematically, num = num + 1
does not make sense. In computing, however, it is used to increment the value of a given variable
by one. It is more appropriate, therefore, to think of the = symbol as an arrow symbol, as shown in
Figure below

![image.png](attachment:image.png)

When thought of this way, it makes clear that the right side of an assignment is evaluated first, then
the result is assigned to the variable on the left . An arrow symbol is not used simply because there
is no such character on a standard computer keyboard. Variables may also be assigned to the value
of another variable (or expression, discussed below) as depicted in Figure below

![image.png](attachment:image.png)

Variables num and k are both associated with the same literal value 10 in memory. One way to see
this is by use of built-in function id,

In [0]:
num=10
id(num)

140726592925728

In [0]:
num=10
k=num
id(k)

140726592925728

The id function produces a unique number identifying a specific value (object) in memory. Since
variables are meant to be distinct, it would appear that this sharing of values would cause problems.
Specifically, if the value of num changed, would variable k change along with it? This cannot happen
in this case because the variables refer to integer values, and integer values are immutable . An
immutable value is a value that cannot be changed. Thus, both will continue to refer to the same
value until one (or both) of them is reassigned, as depicted in Figure below

![image.png](attachment:image.png)

If no other variable references the memory location of the original value, the memory location is
deallocated (that is, it is made available for reuse).

Finally, in Python the same variable can be associated with values of different type during
program execution, as indicated below.

var = 12                                                  integer<br>
var = 12.45                                               float<br>
var = 'Hello'                                             string

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.

In [0]:
num = 10 
num

10

In [0]:
id(num)

140726592925728

In [0]:
id(k)

140726592925728

In [0]:
num = 20
num

20

In [0]:
id(num)

140726592926048

In [0]:
 k
k = num  
k

20

In [0]:
id(k)


140726592926048

In [0]:
id(num)

140726592926048

In [0]:
k = 30
k


30

In [0]:
num

20

In [0]:
id(num)

140726592926048

In [0]:
k = k + 1

In [0]:
id(num)

140726592926048

In [0]:
id(k)

140726592926400

A variable is a name that is associated with a value. The assignment operator, =, is used to
assign values to variables. An immutable value is a value that cannot be changed.

## Variable Assignment and Keyboard Input
The value that is assigned to a given variable does not have to be specified in the program, as demonstrated
in previous examples. The value can come from the user by use of the input function
introduced in Chapter 1,

In [0]:
name = input('What is your first name?')

What is your first name?John


In this case, the variable name is assigned the string 'John'. If the user hit return without entering
any value, name would be assigned to the empty string (' ').

All input is returned by the input function as a string type. For the input of numeric values,
the response must be converted to the appropriate type. Python provides built-in type conversion
functions int () and float () for this purpose, as shown below for a gpa calculation program,

In [0]:
line = input('How many credits do you have?')
num_credits = int(line)
line = input('What is your grade point average?')
gpa = float(line)

How many credits do you have?10
What is your grade point average?4


Here, the entered number of credits, say '9', is converted to the equivalent integer value, 9,
before being assigned to variable num_credits. For input of the gpa, the entered value,

say '3.2', is converted to the equivalent floating-point value, 3.2. Note that the program lines
above could be combined as follows,

In [0]:
num_credits = int(input('How many credits do you have? '))
gpa = float(input('What is your grade point average? '))

How many credits do you have? 10
What is your grade point average? 4


## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results

In [0]:
num = input('Enter number: ')

Enter number: 100


In [0]:
num = int(input('Enter number: '))

Enter number: 100


In [0]:
num = input('Enter name: ')

Enter name: Yash


In [0]:
#Observe error
num = int(input('Enter name: '))

Enter name: abc


ValueError: invalid literal for int() with base 10: 'abc'

All input is returned by the input function as a string type. Built-in functions int() and
float() can be used to convert a string to a numeric type.

### What Is an Identifier?
An identifier is a sequence of one or more characters used to provide a name for a given program
element. Variable names line, num_credits, and gpa are each identifiers. Python is case
sensitive , thus, Line is different from line. Identifiers may contain letters and digits, but cannot
begin with a digit. The underscore character, _, is also allowed to aid in the readability of long identifier names. It should not be used as the first character, however, as identifiers beginning with an underscore have special meaning in Python.

Spaces are not allowed as part of an identifier. This is a common error since some operating
systems allow spaces within file names. In programming languages, however, spaces are used to
delimit (separate) distinct syntactic entities. Thus, any identifier containing a space character
would be considered two separate identifiers. Examples of valid and invalid identifiers in Python
are given in Figure below

![image.png](attachment:image.png)

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.

In [0]:
#Valid identifier
spring2014SemCredits = 15

In [0]:
#Valid identifier
spring2014_sem_credits = 15


In [0]:
#Observe the error
spring2014-sem-credits = 15

SyntaxError: can't assign to operator (<ipython-input-103-63fb3fab2e33>, line 2)

In [0]:
#Observe the error
2014SpringSemesterCredits = 15

SyntaxError: invalid syntax (<ipython-input-104-40a631089339>, line 2)

An identifier is a sequence of one or more characters used to name a given program element. In
Python, an identifier may contain letters and digits, but cannot begin with a digit. The special
underscore character can also be used.

## Keywords and Other Predefined Identifiers in Python
A keyword is an identifier that has predefined meaning in a programming language. Therefore,
keywords cannot be used as “regular” identifiers. Doing so will result in a syntax error, as demonstrated
in the attempted assignment to keyword and below,

In [0]:
#Observe the error
and = 10

SyntaxError: invalid syntax (<ipython-input-105-33ca2c745240>, line 2)

The keywords in Python are listed in Figure below To display the keywords, type help() in the
Python shell, and then type keywords (type 'q' to quit).

![image.png](attachment:image.png)

There are other predefined identifiers that can be used as regular identifiers, but should not be. This
includes float, int, print, exit, and quit, for example. A simple way to check whether
a given identifier is a keyword in Python is given below

In [0]:
'exit' in dir(__builtins__)

False

In [0]:
'exit_program' in dir(__builtins__)

False

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.

In [0]:
#observe the error
yield = 1000

SyntaxError: invalid syntax (<ipython-input-108-238b27284d87>, line 2)

In [0]:
Yield = 1000

In [0]:
print('Hello')

Hello


print = 10

#Observe the error
print('Hello')

A keyword is an identifier that has predefined meaning in a programming language and therefore
cannot be used as a “regular” identifier. Doing so will result in a syntax error.

## Let’s Apply It—“Restaurant Tab Calculation”
The program below calculates a restaurant tab for a couple based on the use of a gift certificate and
the items ordered. This program utilizes the following programming features:

➤ variables<br>
➤ keyboard input<br>
➤ built-in format function<br>
➤ type conversion functions<br>

An example execution of the program is given in Figure below

![image.png](attachment:image.png)

The program is given in Figure below. Lines 1–2 contain comment lines describing what the program
does. The remaining comment lines provide an outline of the basic program sections. Line 5 provides
the required initialization of variables in the program, with variable tax assigned to 8% (.08).
Variable tax is used throughout the program (in lines 9, 35, and 39 ). Thus, if the restaurant tax needs
to be altered, only this line of the program needs to be changed. 

Lines 8–9 display to the user what the program does. The control character \n as the last
character of the print function causes a screen line to be skipped before the next line is displayed.
The cost of the menu items ordered is obtained from the user in lines 15–27 .

Lines 30 and 31 total the cost of the orders for each person, assigned to variables amt_person1
and amt_person2. Lines 34 and 35 compute the tab, including tax (stored in variable tab). Finally,
lines 38–41 display the cost of the ordered items, followed by the added restaurant tax and the amount
due after deducting the amount of the gift certificate. The customers owe any remaining amount

In [0]:
# Restaurant  Tab   Calculation  Program
# This  program will  ca1culate  a  restaurant   tab  with  a gift  certificate:

# initialization
tax = 0.08

# program  greeting
print('This program  will  calculate  a  restaurant tab  for:  a  couple  with')
print('a gift certificate  with  a  restaurant  tax of', tax * 100, '%\n')

# get  amount  of  gift  certificate
amt_certificate  = float(input('Enter amount of the gift certificate:'))

# cost of ordered items
print('Enter ordered items for person 1')

appetizer_per1 = float(input('Appetizier:'))
entree_per1 = float(input('Entree:'))
drinks_per1 = float(input('Drinks:'))
dessert_per1 = float(input('Dessert:'))

print('\nEnter ordered items for person 2')

appetizer_per2 = float(input('Appetizier:'))
entree_per2 = float(input('Entree:'))
drinks_per2 = float(input('Drinks:'))
dessert_per2 = float(input('Dessert:'))

# total items
amt_person1 = appetizer_per1 + entree_per1 + drinks_per1 + dessert_per1
amt_person2 = appetizer_per2 + entree_per2 + drinks_per2 + dessert_per2

# compute tab with tax
items_cost = amt_person1 + amt_person2
tab = items_cost + items_cost * tax

# display amount  owe
print('\nOdered items: $', format(items_cost,'.2f'))
print('Restaurant tax: $', format(items_cost * tax, '.2f'))
print('Tab: $', format(tab - amt_certificate,'.2f'))
print('(negative amount indicates unused amount of gift certificate)')

This program  will  calculate  a  restaurant tab  for:  a  couple  with
a gift certificate  with  a  restaurant  tax of 8.0 %

Enter amount of the gift certificate:200
Enter ordered items for person 1
Appetizier:1
Entree:21.50
Drinks:4.25
Dessert:6

Enter ordered items for person 2
Appetizier:7.34
Entree:2.33
Drinks:7
Dessert:5

Odered items: $ 54.42
Restaurant tax: $ 4.35
Tab: $ -141.23
(negative amount indicates unused amount of gift certificate)


A negative amount indicates the amount left on the gift certificate. Built-in function format is used to
limit the output to two decimal places.

## Operators
Now that we have used numeric and string types in Python, we look at operations that may be performed
on them.

## What Is an Operator?
An operator is a symbol that represents an operation that may be performed on one or more operands
. For example, the + symbol represents the operation of addition. An operand is a value that a
given operator is applied to, such as operands 2 and 3 in the expression 2 + 3. A unary operator
operates on only one operand, such as the negation operator in - 12. A binary operator operates on
two operands, as with the addition operator. Most operators in programming languages are binary
operators. We look at the arithmetic operators in Python next

An operator is a symbol that represents an operation that may be performed on one or more
operands. Operators that take one operand are called unary operators. Operators that take two
operands are called binary operators.

## Arithmetic Operators
Python provides the arithmetic operators given in Figure below

The + , - , * (multiplication) and / (division) arithmetic operators perform the usual operations.
Note that the - symbol is used both as a unary operator (for negation) and a binary operator
(for subtraction).

20 - 5 ➝ 15 ( - as binary operator)<br>
-10 * 2 ➝ - 20 ( - as unary operator)

![image.png](attachment:image.png)

Python also includes an exponentiation (**) operator. Integer and floating-point values can be used
in both the base and the exponent

2**4 ➝ 16<br>
2.5 ** 4.5 ➝ 61.76323555016366

Python provides two forms of division. “true” division is denoted by a single slash, /. Thus, 25 /
10 evaluates to 2.5. Truncating division is denoted by a double slash, //, providing a truncated
result based on the type of operands applied to. When both operands are integer values, the result is a
truncated integer referred to as integer division . When as least one of the operands is a fl oat type, the
result is a truncated floating point. Thus, 25 // 10 evaluates to 2, while 25.0 // 10 becomes
2.0. This is summarized in Figure below

![image.png](attachment:image.png)

An example of the use of integer division would be to determine the number of dozen doughnuts for
a given number of doughnuts. If variable numDoughnuts had a current value of 29, the number
of dozen doughnuts would be calculated by,

numDoughnuts // 12 ➝ 29 // 12 ➝ 2

Lastly, the modulus operator (%) gives the remainder of the division of its operands, resulting in
a cycle of values. This is shown in Figure below

![image.png](attachment:image.png)

The modulus and truncating (integer) division operators are complements of each other. For example,
29 // 12 gives the number of dozen doughnuts, while 29 % 12 gives the number of leftover
doughnuts (5).

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.

In [0]:
10 + 35

45

In [0]:
-10 + 35

25

In [0]:
4 * 2

8

In [0]:
4 ** 2

16

In [0]:
45 / 10

4.5

In [0]:
45 // 10

4

In [0]:
45 // 10.0

4.0

In [0]:
2025 % 10

5

In [0]:
2025 // 10

202

The division operator, /, produces “true division” regardless of its operand types. The truncating
division operator, //, produces either an integer or float truncated result based on the type
of operands applied to. The modulus operator (%) gives the remainder of the division of its
operands.

## Let’s Apply It—“Your Place in the Universe”
The following program (Figure below) calculates the approximate number of atoms that the average
person contains, and the percentage of the universe that they comprise. This program utilizes the
following programming features:

➤ floating-point scientific notation   ➤ built-in format function

In [0]:
#Program Execution......
#This program will detrmine your place in the universe.
#Enter your weight in ponds:150
    #You contain approximately 3.30e+28 atoms
    #Therefore, you comprise 3.30e-51 % of the universe
    
    # Your place in the Universe Program
    # This program will determine the approximate number of atoms that are a
    # person consists of and the percent of the universe that they comprise.
    
    #initialization
num_atoms_universe = 10e80
weight_avg_person = 70 # 70 kg (154 lbs)
num_atoms_avg_person = 7e27
    
 # program greeting
print('This program will determine your place in the universe.')
    
    # prompt for user's weight:
weight_lbs = int(input('Enter your weight in ponds:'))
    
    # convert weight to kilograms
weight_kg = 2.2 * weight_lbs
    
    # determine number atoms in person
num_atoms = (weight_kg / 70) * num_atoms_avg_person
percent_of_universe = (num_atoms / num_atoms_universe) * 100
    
    # display results
print('You contain approximately', format(num_atoms,'2e'),'atoms')
print('Therefore, you comprise', format(percent_of_universe, '.2e'),'% of the univese')
    

This program will determine your place in the universe.
Enter your weight in ponds:150
You contain approximately 3.300000e+28 atoms
Therefore, you comprise 3.30e-51 % of the univese


Lines 1–4 describe the program. Needed variables num_atoms_universe, weight_avg_
person, and num_atoms_avg_person are initialized in lines 7–9 . The program greeting is on
line 12 . Line 15 inputs the person’s weight. Line 18 converts the weight to kilograms for use in the
calculations on lines 21–22 which compute the desired results. Finally, lines 25–27 display the results

## Expressions and Data Types
Now that we have looked at arithmetic operators, we will see how operators and operands can be
combined to form expressions. In particular, we will look at how arithmetic expressions are evaluated
in Python. We also introduce the notion of a data type .

## What Is an Expression?
An expression is a combination of symbols that evaluates to a value. Expressions, most commonly,
consist of a combination of operators and operands,

4 + (3 * k)

An expression can also consist of a single literal or variable. Thus, 4, 3, and k are each expressions.
This expression has two subexpressions, 4 and (3 * k). Subexpression (3 * k) itself has two
subexpressions, 3 and k.

Expressions that evaluate to a numeric type are called arithmetic expressions . A subexpression
is any expression that is part of a larger expression. Subexpressions may be denoted by the use
of parentheses, as shown above. Thus, for the expression 4 + (3 * 2), the two operands of the
addition operator are 4 and (3 * 2), and thus the result is equal to 10. If the expression were
instead written as (4 + 3) * 2, then it would evaluate to 14.

Since a subexpression is an expression, any subexpression may contain subexpressions of its own,

In [0]:
4 + (3 * (2 - 1))     #➝ 4 + (3 * 1) ➝ 4 + 3 ➝ 7

7

If no parentheses are used, then an expression is evaluated according to the rules of operator precedence
in Python, discussed in the next section.

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.

In [0]:
(2 + 3) * 4

20

In [0]:
2 + (3 * 4)

14

In [0]:
2 + ((3 * 4) - 8)

6

In [0]:
2 + 3 * (4 - 1)

11

An expression is a combination of symbols (or single symbol) that evaluates to a value. A subexpression
is any expression that is part of a larger expression.

## Operator Precedence
The way we commonly represent expressions, in which operators appear between their operands,
is referred to as infix notation . For example, the expression 4 + 3 is in infix notation since
the + operator appears between its two operands, 4 and 3. There are other ways of representing
expressions called prefix and postfix notation, in which operators are placed before and after their
operands, respectively.

The expression 4 + (3 * 5) is also in infix notation. It contains two operators, + and *.
The parentheses denote that (3 * 5) is a subexpression. Therefore, 4 and (3 * 5) are the
operands of the addition operator, and thus the overall expression evaluates to 19. What if the parentheses
were omitted, as given below?

4 + 3 * 5

How would this be evaluated? These are two possibilities,

In [0]:
4 + 3 * 5       # ➝ 4 + 15 ➝ 19<br>


19

In [0]:
4 + 3 * 5       #➝ 7 * 5 ➝ 35

19

Some might say that the first version is the correct one by the conventions of mathematics. However,
each programming language has its own rules for the order that operators are applied, called operator
precedence , defined in an operator precedence table . This may or may not be the same as in
mathematics, although it typically is. In Figure below, we give the operator precedence table for the
Python operators discussed so far. (We will discuss the issue of associativity indicated in Figure  below
in the next section.)

![image.png](attachment:image.png)

In the table, higher-priority operators are placed above lower-priority ones. Thus, we see that multiplication
is performed before addition when no parentheses are included,

In [0]:
4 + 3 * 5 

19

In our example, therefore, if the addition is to be performed fi rst, parentheses would be needed,

In [0]:
(4 + 3) * 5 

35

As another example, consider the expression below. Following Python’s rules of operator precedence,
the exponentiation operator is applied first, then the truncating division operator, and finally
the addition operator,

In [0]:
4 + 2 ** 5 // 10 

7

Operator precedence guarantees a consistent interpretation of expressions. However, it is good programming
practice to use parentheses even when not needed if it adds clarity and enhances readability,
without overdoing it. Thus, the previous expression would be better written as,

In [0]:
4 + (2 ** 5) // 10

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.

In [0]:
2 + 3 * 4

14

In [0]:
2 * 3 + 4

10

In [0]:
2 * 3 / 4

1.5

In [0]:
2 * 3 // 4

1

In [0]:
5 + 42 % 10

7

In [0]:
2 * 2 ** 3

16

Operator precedence is the relative order that operators are applied in the evaluation of
expressions, defined by a given operator precedence table.

## Operator Associativity
A question that you may have already had is, “What if two operators have the same level of precedence,
which one is applied first?” For operators following the associative law, the order of evaluation
doesn’t matter,

(2 + 3) + 4 ➝ 9        <br>                             2 + (3 + 4) ➝ 9

In this case, we get the same results regardless of the order that the operators are applied. 

In [0]:
(8-4)-2

2

In [0]:
8-(4-2)

6

Division and subtraction, however, do not follow the associative law.

In [0]:
(8 / 4) / 2

1.0

In [0]:
8 / (4 / 2)

4.0

In [0]:
2 ** (3 ** 2)

512

In [0]:
(2 ** 3) ** 2

64

Here, the order of evaluation does matter. To resolve the ambiguity, each operator has a specified
operator associativity that defines the order that it and other operators with the same level of precedence
are applied. All operators in the figure, except for exponentiation,
have left-to-right associativity—exponentiation has right-to-left associativity

## L E T ’ S T R Y I T
From the Python Shell, enter the following and observe the results.

In [0]:
6 - 3 + 2

5

In [0]:
(6 - 3) + 2

5

In [0]:
6 - (3 + 2)

1

In [0]:
2 * 3 / 4

1.5

In [0]:
12 % (10 / 2)

2.0

In [0]:
2 ** 2 ** 3

256

In [0]:
(2 ** 2) ** 3

64

In [0]:
2 ** (2 ** 3)

256

Operator associativity is the order that operators are applied when having the same level of
precedence, specific to each operator.

## What Is a Data Type?
A data type is a set of values, and a set of operators that may be applied to those values. For
example, the integer data type consists of the set of integers, and operators for addition, subtraction,
multiplication, and division, among others. Integers, floats, and strings are part of a set of predefined
data types in Python called the built-in types .

Data types prevent the programmer from using values inappropriately. For example, it does
not make sense to try to divide a string by two, 'Hello' / 2. The programmer knows this by
common sense. Python knows it because 'Hello' belongs to the string data type, which does not
include the division operation. The need for data types results from the fact that the same internal
representation of data can be interpreted in various ways.

The sequence of bits in the figure can be interpreted as a character ('A') or an integer (65). If a
programming language did not keep track of the intended type of each value, then the programmer
would have to. This would likely lead to undetected programming errors, and would provide even
more work for the programmer. We discuss this further in the following section.

Finally, there are two approaches to data typing in programming languages. In static typing ,
a variable is declared as a certain type before it is used, and can only be assigned values of that type.
Python, however, uses dynamic typing . In dynamic typing , the data type of a variable depends only
on the type of value that the variable is currently holding. Thus, the same variable may be assigned
values of different type during the execution of a program.

A data type is a set of values, and a set of operators that may be applied to those values.

## Mixed-Type Expressions
A mixed-type expression is an expression containing operands of different type. The CPU can only
perform operations on values with the same internal representation scheme, and thus only on operands
of the same type. Operands of mixed-type expressions therefore must be converted to a common
type. Values can be converted in one of two ways—by implicit (automatic) conversion, called
coercion , or by explicit type conversion . We look at each of these next.

A mixed-type expression is an expression with operands of different type.

Mixed-Type Expressions
A mixed-type expression is an expression containing operands of different type. The CPU can only
perform operations on values with the same internal representation scheme, and thus only on operands
of the same type. Operands of mixed-type expressions therefore must be converted to a common
type. Values can be converted in one of two ways—by implicit (automatic) conversion, called
coercion , or by explicit type conversion . We look at each of these next

A mixed-type expression is an expression with operands of different type.

## Coercion vs. Type Conversion
Coercion is the implicit (automatic) conversion of operands to a common type. Coercion is automatically
performed on mixed-type expressions only if the operands can be safely converted,
that is, if no loss of information will result. The conversion of integer 2 to floating-point 2.0
below is a safe conversion—the conversion of 4.5 to integer 4 is not, since the decimal digit
would be lost,

In [0]:
2 + 4.5 #➝ 2.0 + 4.5 ➝ 6.5 safe (automatic conversion of int to float)

6.5

Type conversion is the explicit conversion of operands to a specific type. Type conversion can be
applied even if loss of information results. Python provides built-in type conversion functions
int() and float(), with the int() function truncating results as given in Figure below

In [0]:
float(2) + 4.5   #➝ 2.0 + 4.5 ➝ 6.5


6.5

In [0]:
2 + int(4.5)    # ➝ 2 +4 ➝ 6

6

![image.png](attachment:image.png)

Note that numeric strings can also be converted to a numeric type. In fact, we have already been
doing this when using int or fl oat with the input function,

In [0]:
num_credits = int(input('How many credits do you have? '))

How many credits do you have? 300


Coercion is the implicit (automatic) conversion of operands to a common type. Type conversion
is the explicit conversion of operands to a specific type

## Let’s Apply It—“Temperature Conversion Program”
The following Python program (Figure below) requests from the user a temperature in degrees
Fahrenheit, and displays the equivalent temperature in degrees Celsius. This program utilizes the
following programming features:

➤ arithmetic expressions   ➤ operator associativity   ➤ format function

In [0]:
#This program will convert degrees Fahrenheit to degrees celsius
#Enter degrees Fahrenheit: 100
#100.0 degrees Farenheit equals 37.8 degree celsius

#Temperature conversion program (Fahrenheit to celsius)
#program greeting
print('This program will convert degrees Fahrenheit to degrees Celsius')

#get temperature in Fahrenheit
fahren = float(input('Enter degrees Fahrenheit:'))
#calc degrees celsius
celsius = (fahren - 32) * 5 / 9


# output degrees celsius
print(fahren, 'degrees Fahrenheit equals',
      format (celsius,'.1f'),'derees celsius')

This program will convert degrees Fahrenheit to degrees Celsius
Enter degrees Fahrenheit:244
244.0 degrees Fahrenheit equals 117.8 derees celsius
